techmatters/hrm
# HRM support for async jobs II: Create and process transcript jobs

Primary reviewer: @stephenhand 

## Description
This PR:
- Adds `mediaUrls?: ContactMediaUrl[]` property to `Contact`'s JSON column, and `appendMediaUrls` function to contacts data access to append urls to it.
- Moves logic to make contact creation idempotent from data access layer to business layer. Can revert this change if not appropriate, it just felt a better fit for this code (cause it's related to how this entities behaves).
- Re-exports things coming from contact data access layer in contact business layer, so other modules can consume from there (feels more appropriate to me? Happy to rollback if this change is not desired).
- Adds support for the `retrieve-transcript` contact job type. This includes
  - Logic to create a pending job when a chat contact is submitted.
  - Logic to publish due jobs of this type into the via the publish adapter.
  - Logic to process the jobs of this type when polled.
- Types fixes and adjustments.
- Adds unit tests and service tests around the jobs system.

Notes:
- In a following PR I'll be adding logic to handle "failed job attempts" from SQS completed queue.

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-1382)

### Verification steps
It will be very helpful to add some debug logs to the jobs processor before this.
- Start a local server, expose via ngrok, point to it from Flex and save a chat contact.
- Confirm that a new pending job was created, and that it is published on the next iteration of the processor.
- Stop the server and "mock" the polling function to return the same contact job as before, with the same shape as it was published, plus a completion payload.
- Run the server again and check that the "faked completed job" is processed accordingly, marking it as completed and adding the given resource url to the actual contact resource.
